src/main/kotlin/fi/metatavu/timebank/api/severa/SeveraAccessTokenContainer.kt

```diff
@@ -0,0 +1,146 @@
+package fi.metatavu.timebank.api.severa
+
+import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
+import fi.metatavu.timebank.api.severa.models.SeveraAccessToken
+import okhttp3.FormBody
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okhttp3.Response
+import org.eclipse.microprofile.config.inject.ConfigProperty
+import org.slf4j.Logger
+import javax.enterprise.context.ApplicationScoped
+import javax.inject.Inject
+
+@ApplicationScoped
+class SeveraAccessTokenContainer {
+
+    @ConfigProperty(name = "severa.demo.base.url")
+    lateinit var severaBaseUrl: String
+
+    @ConfigProperty(name = "severa.demo.client.id")
+    lateinit var severaClientId: String
+
+    @ConfigProperty(name = "severa.demo.client.secret")
+    lateinit var severaClientSecret: String
+
+    @Inject
+    lateinit var logger: Logger
+
+    private var severaAccessToken: SeveraAccessToken? = null
+
+    /**
+     * Determines if a new access token is needed. Returns a functional access token for Severa API calls
+     *
+     * @param scopes List of scopes
+     * @return Access token for Severa API calls
+     */
+    fun getAccessToken(scopes: List<String>): SeveraAccessToken {
+        if (severaAccessToken == null){
+            return getNewAccessToken(scopesToString(scopes))
+        }
+
+        return if (isValidToken(severaAccessToken!!.accessToken) && containsScopes(scopes)){
+            severaAccessToken!!
+        } else {
+            getNewAccessToken(scopesToString(scopes))
+        }
+    }
+
+    /**
+     * Converts a list of scopes to be a suitable string for requesting an access token
+     *
+     * @param scopes List of scopes
+     * @return String
+     */
+    private fun scopesToString(scopes: List<String>): String{
+        val scopeBuilder = StringBuilder()
+        for (scope in scopes){
+            scopeBuilder.append(scope).append(", ")
+        }
+
+        return scopeBuilder.toString()
+    }
+
+    /**
+     * Checks if potentially existing access token has the needed scope(s)
+     *
+     * @param scopes List of scopes
+     * @return Boolean
+     */
+    private fun containsScopes(scopes: List<String>): Boolean {
+        for (scope in scopes){
+            if (!severaAccessToken!!.scope.contains(scope)) return false
+        }
+        return true
+    }
+
+    /**
+     * Checks if an access token is expired. Does NOT check for suitable scopes.
+     *
+     * @param token accessToken
+     * @return boolean
+     */
+    private fun isValidToken(token: String): Boolean{
```

Comment:
This seems like a good idea, but cannot we just determine it from the expiration of the access token?

Reply:
Sure we could, but I don't see why we wouldn't utilize a feature provided by the Severa API? They both have the same end result after all.